appbuilder/services/npm-service.ts

```diff
@@ -0,0 +1,233 @@
+			}
+
+			if (packageJsonContent && packageJsonContent.devDependencies && packageJsonContent.devDependencies[`${NpmService.TYPES_DIRECTORY}${dependency}`]) {
+				this.npmUninstall(projectDir, `${NpmService.TYPES_DIRECTORY}${dependency}`, ["--save-dev"]).wait();
+			}
+		}).future<void>()();
+	}
+
+	public getPackageJsonFromNpmRegistry(packageName: string, version?: string): IFuture<any> {
+		return (() => {
+			let packageJsonContent: any;
+			version = version || "latest";
+			try {
+				let url = this.buildNpmRegistryUrl(packageName, version);
+				// This call will return error with message '{}' in case there's no such package.
+				let result = this.$httpClient.httpRequest(url).wait().body;
+				packageJsonContent = JSON.parse(result);
+			} catch (err) {
+				this.$logger.trace("Error caught while checking the NPM Registry for plugin with id: %s", packageName);
+				this.$logger.trace(err.message);
+			}
+
+			return packageJsonContent;
+		}).future<any>()();
+	}
+
+	private hasTypesForDependency(packageName: string): IFuture<boolean> {
+		return (() => {
+			return !!this.getPackageJsonFromNpmRegistry(`${NpmService.TYPES_DIRECTORY}${packageName}`).wait();
+		}).future<boolean>()();
+	}
+
+	private buildNpmRegistryUrl(packageName: string, version: string): string {
+		return `${NpmService.NPM_REGISTRY_URL}/${packageName.replace("/", "%2F")}?version=${encodeURIComponent(version)}`;
+	}
+
+	private getPackageJsonContent(projectDir: string): IFuture<any> {
+		return (() => {
+			let pathToPackageJson = this.getPathToPackageJson(projectDir);
+
+			try {
+				return this.$fs.readJson(pathToPackageJson).wait();
+			} catch (err) {
+				if (err.code === "ENOENT") {
+					this.$errors.failWithoutHelp(`Unable to find ${this.$projectConstants.PACKAGE_JSON_NAME} in ${projectDir}.`);
+				}
+
+				throw err;
+			}
+
+		}).future<any>()();
+	}
+
+	private getPathToPackageJson(projectDir: string): string {
+		return path.join(projectDir, this.$projectConstants.PACKAGE_JSON_NAME);
+	}
+
+	private getPathToReferencesFile(projectDir: string): string {
+		return path.join(projectDir, this.$projectConstants.REFERENCES_FILE_NAME);
+	}
+
+	private installTypingsForDependency(projectDir: string, dependency: string): IFuture<void> {
+		return this.npmInstall(projectDir, `${NpmService.TYPES_DIRECTORY}${dependency}`, null, ["--save-dev", "--save-exact"]);
+	}
+
+	private generateReferencesFile(projectDir: string): IFuture<void> {
+		return (() => {
+			let packageJsonContent = this.getPackageJsonContent(projectDir).wait();
+
+			let pathToReferenceFile = this.getPathToReferencesFile(projectDir),
+				lines: string[] = [];
+
+			if (packageJsonContent && packageJsonContent.dependencies && packageJsonContent.dependencies[constants.TNS_CORE_MODULES]) {
+				let relativePathToTnsCoreModulesDts = `./${constants.NODE_MODULES_DIR_NAME}/${constants.TNS_CORE_MODULES}/${NpmService.TNS_CORE_MODULES_DEFINITION_FILE_NAME}`;
```

Comment:
Nope, the path in references.d.ts must be Unix style. So I prefer to write it Unix style instead of using `path.join` and convert from Windows style to Unix